mmengine/_strategy/single_device.py

```diff
@@ -0,0 +1,242 @@
+
+    def wrap_model(self, model: nn.Module) -> nn.Module:
+        model = self.convert_model(model)
+        current_device = get_device()
+        return model.to(current_device)
+
+    def load_checkpoint(
+        self,
+        filename: str,
+        *,
+        map_location: Union[str, Callable] = 'cpu',
+        callback: Optional[Callable] = None,
+    ) -> dict:
+        """Load checkpoint from given ``filename``.
+
+        Args:
+            filename (str): Accept local filepath, URL, ``torchvision://xxx``,
+                ``open-mmlab://xxx``.
+            map_location (str or callable): A string or a callable function to
+                specifying how to remap storage locations.
+                Defaults to 'cpu'.
+        """
+        from mmengine.runner.checkpoint import _load_checkpoint
+
+        if hasattr(self, 'extra_ckpt'):
+            return self.extra_ckpt
+
+        self.logger.info(f'Load checkpoint from {filename}')
+
+        self.extra_ckpt: dict
+        if map_location == 'default':
+            device = get_device()
+            self.extra_ckpt = _load_checkpoint(filename, map_location=device)
+        else:
+            self.extra_ckpt = _load_checkpoint(
+                filename, map_location=map_location)
+
+        # users can do some modification after loading checkpoint
+        if callback is not None:
+            callback(self.extra_ckpt)
+
+        state_dict = self.extra_ckpt.pop('state_dict')
+        self.load_model_state_dict(state_dict)
+
+        return self.extra_ckpt
+
+    def resume(
+        self,
+        filename: str,
+        *,
+        resume_optimizer: bool = True,
+        resume_param_scheduler: bool = True,
+        map_location: Union[str, Callable] = 'default',
+        callback: Optional[Callable] = None,
+    ) -> dict:
+        """Resume training from given ``filename``.
+
+        Four types of states will be resumed.
+
+        - model state
+        - optimizer state
+        - scheduler state
+        - randomness state
+
+        Args:
+            filename (str): Accept local filepath, URL, ``torchvision://xxx``,
+                ``open-mmlab://xxx``.
+        """
+        if hasattr(self, 'extra_ckpt'):
+            return self.extra_ckpt
+
+        self.logger.info(f'Resume checkpoint from {filename}')
+
+        self.extra_ckpt = self.load_checkpoint(
+            filename, map_location=map_location, callback=callback)
+
+        if not resume_optimizer:
+            self.extra_ckpt.pop('optimizer', None)
+        else:
+            self.load_optim_state_dict(self.extra_ckpt.pop('optimizer'))
+
+        if not resume_param_scheduler:
+            self.extra_ckpt.pop('param_schedulers', None)
+        else:
+            self.load_scheduler_state_dict(
+                self.extra_ckpt.pop('param_schedulers'))
+
+        # resume random seed
+        resumed_seed = self.extra_ckpt['meta'].get('seed', None)
```

Comment:
Should we integrate `MessageHub` in strategy or still use `meta`

Reply:
`meta` is an outdated design for `Runner`, but we still keep this field for avoiding BC-breaking. I think using `MessageHub` will not make `Strategy` harder to use.

Reply:
I want the strategy to be simpler, with fewer concepts that are not related to native PyTorch. If the user needs to record information and other operations, they need to handle it themselves.